File: leetcode/plus_one/PlusOne.cpp
class Solution {
public:
    vector<int> plusOne(vector<int> &digits) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        vector<int> result; 
        if (digits.size() == 0) {
            return result; 
        }
        
        int carry = 1; 
        
        for (int i = digits.size() - 1; i >= 0; i--) {
            int digit = (digits[i] + carry) % 10; 
            carry = (digits[i] + carry) / 10; 
            result.insert(result.begin(), digit); 
        }
        
        if (carry == 1) {
            result.insert(result.begin(), 1); 
        }
        
        return result; 
    }
};

File: leetcode4/PlusOne.java
public class Solution {
    public int[] plusOne(int[] digits) {
        int len = digits.length; 
        int carry = 1; 
        
        for (int i = len - 1; i >= 0; i--) {
            int sum = digits[i] + carry; 
            int digit = sum % 10; 
            carry = sum / 10; 
            digits[i] = digit; 
        }
        
        if (carry == 0) {
            return digits; 
        }
        
        int[] temp = new int[len + 1];
        temp[0] = carry; 
        for (int i = 0; i < len; i++) {
            temp[i + 1] = digits[i]; 
        }
        
        return temp; 
    }
}
